larq_compute_engine/mlir/transforms/optimize_patterns.td

```diff
@@ -12,96 +12,148 @@ def HasOneUse : Constraint<CPred<"$0.hasOneUse()">>;
-                        (binaryOp (ConstantOp $post_activation_multiplier),
-                                  (ConstantOp $value), TFL_AF_None),
-                        (binaryOp (ConstantOp $post_activation_bias),
-                                  (ConstantOp $value), TFL_AF_None),
-                        $channels_in, $dilation_height_factor, $dilation_width_factor,
-                        $fused_activation_function,
-                        $pad_values, $padding, $stride_height, $stride_width),
-         [(HasOneUse $conv_output)], (addBenefit 100)>;
+  def : Pat<(binaryOp
+                (TF_LceBconv2dOp:$conv_output
+                    $input,
+                    $filter,
+                    (ConstantOp F32ElementsAttr:$post_activation_multiplier),
+                    (ConstantOp F32ElementsAttr:$post_activation_bias),
+                    $write_bitpacked_threshold,
+                    $channels_in,
+                    $dilation_height_factor,
+                    $dilation_width_factor,
+                    $fused_activation_function,
+                    $pad_values,
+                    $padding,
+                    $stride_height,
+                    $stride_width),
+                (ConstantOp F32ElementsAttr:$value), TFL_AF_None),
+            (TF_LceBconv2dOp
+                $input,
+                $filter,
+                (binaryOp (ConstantOp $post_activation_multiplier),
+                          (ConstantOp $value), TFL_AF_None),
+                (binaryOp (ConstantOp $post_activation_bias),
+                          (ConstantOp $value), TFL_AF_None),
+                $write_bitpacked_threshold,
+                $channels_in,
+                $dilation_height_factor,
+                $dilation_width_factor,
+                $fused_activation_function,
+                $pad_values,
+                $padding,
+                $stride_height,
+                $stride_width),
+            [(HasOneUse $conv_output)], (addBenefit 100)>;
 }
 
-def HasNegativeValues : AttrConstraint<CPred<"HasNegativeValues($_self)">>;
-def ComputeBSignAndExpandTo4D : NativeCodeCall<"ComputeBSignAndExpandTo4D($0)">;
-
-def : Pat<(TF_LceBconv2dOp $input, (ConstantOp F32ElementsAttr:$filter),
-            (ConstantOp HasNegativeValues:$post_activation_multiplier),
-            $post_activation_bias, $channels_in,
-            $dilation_height_factor, $dilation_width_factor,
-            TFL_AF_None, $pad_values, $padding,
-            $stride_height, $stride_width),
-          (TF_LceBconv2dOp $input,
-            (TFL_MulOp (ConstantOp $filter), (ConstantOp (ComputeBSignAndExpandTo4D $post_activation_multiplier)), TFL_AF_None),
-            (TFL_AbsOp (ConstantOp $post_activation_multiplier)),
-            $post_activation_bias, $channels_in,
-            $dilation_height_factor, $dilation_width_factor,
-            TFL_AF_None, $pad_values, $padding,
-            $stride_height, $stride_width),
-          [], (addBenefit 90)>;
```

Comment:
Just curious, what is the benefit of moving this transform into the activation bitpacking pattern and not doing this before hand? Is it due to the handling of the fused activation?

Reply:
Yeah, the issue is that we need to know inside the activation bitpacking pass which of the weights have been flipped, because for channels with flipped weights we need to invert the clamp min/max for computing the correct threshold.

Reply:
I see, makes sense :+1: